File: c/tools/prime.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

int main(int argc, char* argv[]) {
  if (argc == 1) {
    printf("Usage ?prime_start=? ?prime_end=? prime [num_primes]\n");
    return 0;
  }

  int num_primes = atoi(argv[1]);
  const char* range_begin = getenv("prime_start");
  const char* range_end = getenv("prime_end");

  int start = range_begin != NULL ? atoi(range_begin) : 1;
  int end = range_end != NULL ? atoi(range_end) : INT_MAX;

  for (int i = start, count = 0; i < end && count < num_primes; i++) {
    int half = i / 2;
    int prime = 1;

    for (int j = 2; j <= half; j++) {
      if (i % j == 0) {
	prime = 0;
	break;
      }
    }

    if (prime) {
      printf("%d\n", i);
      count++;
    }
  }

  return 0;
}

File: ruby/tools/prime.rb
#!/bin/ruby

class Fixnum
  N_BYTES = [42].pack('i').size
  N_BITS = N_BYTES * 8
  MAX = 2 ** (N_BITS - 2) - 1
  MIN = -MAX - 1
end

if ARGV.length == 0 
  puts 'need number of primes'
  exit
end

num_primes = ARGV[1]
range_begin = ENV['prime_start']
range_end = ENV['prime_end']

start = range_begin ? range_begin.to_i : 1
stop = range_end ? range_end.to_i : Fixnum::MAX

i = start
count = 0
while i < stop && count < num_primes do
  half = i / 2

  for j in 2..half
    if i % j == 0
      puts i
      count += 1 
      break
    end
  end

  i++
end

